dali/operators/reader/loader/video_loader.cc

```diff
@@ -211,19 +212,76 @@ typename std::enable_if<!std::numeric_limits<T>::is_integer, bool>::type
         || std::abs(x-y) < std::numeric_limits<T>::min();
 }
 
-OpenFile& VideoLoader::get_or_open_file(const std::string &filename) {
+static int read_packet(void *opaque, uint8_t *buf, int buf_size) {
+  VideoFileDesc *file_data = static_cast<VideoFileDesc*>(opaque);
+
+  // open file if needed
+  if (!file_data->file_stream) {
+    file_data->file_stream = fopen(file_data->filename.c_str(), "r");
+    DALI_ENFORCE(file_data->file_stream, std::string("Could not open file ") + file_data->filename);
+    fseek(file_data->file_stream, file_data->file_position, SEEK_SET);
+  }
+  return fread(buf, 1, buf_size, file_data->file_stream);
+}
+
+static int64_t seek_file(void *opaque, int64_t offset, int whence) {
+  VideoFileDesc *file_data = static_cast<VideoFileDesc*>(opaque);
+
+  // open file if needed
+  if (!file_data->file_stream) {
+    file_data->file_stream = fopen(file_data->filename.c_str(), "r");
+    DALI_ENFORCE(file_data->file_stream, std::string("Could not open file ") + file_data->filename);
+    fseek(file_data->file_stream, file_data->file_position, SEEK_SET);
+  }
+  int ret = -1;
+  switch (whence) {
+    case SEEK_SET:
+    case SEEK_CUR:
+    case SEEK_END:
+      ret = fseek(file_data->file_stream, offset, whence);
+      break;
+    default:
+      break;
+  }
+  if (ret == 0) {
+    return ftell(file_data->file_stream);
+  } else {
+    // fseek may return > 0 but it is still an error
+    return -abs(ret);
+  }
+}
+
+VideoFile& VideoLoader::get_or_open_file(const std::string &filename) {
   auto& file = open_files_[filename];
 
-  if (!file.open) {
+  if (file.empty()) {
+    file.file_desc.filename = filename;
     LOG_LINE << "Opening file " << filename << std::endl;
 
-    AVFormatContext* raw_fmt_ctx = nullptr;
-    int ret = avformat_open_input(&raw_fmt_ctx, filename.c_str(), NULL, NULL);
-    if (ret < 0) {
-      DALI_FAIL(std::string("Could not open file ") + filename + " because of " + av_err2str(ret));
-    }
-    file.fmt_ctx_ = make_unique_av<AVFormatContext>(raw_fmt_ctx, avformat_close_input);
+    auto raw_fmt_ctx = std::unique_ptr<AVFormatContext, void(*)(AVFormatContext*)>(
+                                       avformat_alloc_context(), avformat_free_context);
+
+    DALI_ENFORCE(raw_fmt_ctx, "Failed to allocate avformat_alloc_context");
+
+    size_t avio_ctx_buffer_size = 4096;
+    uint8_t *avio_ctx_buffer = static_cast<uint8_t *>(av_malloc(avio_ctx_buffer_size));
```

Comment:
would be good to wrap this one with a unique_ptr

Reply:
Done